File: src/test/java/org/minturtle/careersupport/auth/utils/ApiTokenProviderTest.java
package org.minturtle.careersupport.auth.utils;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.ValueSource;
import org.minturtle.careersupport.common.exception.BadRequestException;
import org.minturtle.careersupport.common.utils.ReactiveEncryptUtils;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

import static org.assertj.core.api.Assertions.*;

class ApiTokenProviderTest {

    private ReactiveEncryptUtils reactiveEncryptUtils =
            new ReactiveEncryptUtils("X1XnwJ2Vrdw9wqdfX0rOdLfNJ8rwrvB9");
    private ApiTokenProvider apiTokenProvider =
            new ApiTokenProvider(reactiveEncryptUtils);


    @Test
    @DisplayName("API 토큰을 생성할 수 있다.")
    void testGenerateShouldReturnNonEmptyToken() {
        String testUserId = "abc123";
        Mono<String> tokenMono = apiTokenProvider.generate(testUserId);

        StepVerifier.create(tokenMono)
                .assertNext(token -> {
                    assertThat(token).isNotBlank();
                    assertThat(token).startsWith("cs_");
                    assertThat(token.length()).isGreaterThan(3);
                })
                .verifyComplete();
    }

    @Test
    @DisplayName("생성한 토큰을 해독해서 userId를 반환받을 수 있다.")
    void testDecryptApiTokenShouldReturnOriginalUserId() {
        String testUserId = "abc123";
        Mono<String> tokenMono = apiTokenProvider.generate(testUserId);

        StepVerifier.create(tokenMono)
                .assertNext(token -> {
                    Mono<String> decryptedUserIdMono = apiTokenProvider.decryptApiToken(token);

                    StepVerifier.create(decryptedUserIdMono)
                            .expectNext(testUserId)
                            .verifyComplete();
                })
                .verifyComplete();
    }

    @ParameterizedTest
    @ValueSource(strings = {"", " ", "invalid", "cs_invalid"})
    @DisplayName("API 토큰이 잘못된 경우 BadRequestException을 throw한다.")
    public void testDecryptThrowsBadRequestException(String invalidToken) throws Exception{
        Mono<String> decryptedUserIdMono = apiTokenProvider.decryptApiToken(invalidToken);

        StepVerifier.create(decryptedUserIdMono)
                .expectError(BadRequestException.class)
                .verify();

    }

}

File: src/test/kotlin/org/minturtle/careersupport/common/utils/ApiTokenProviderTest.kt
package org.minturtle.careersupport.common.utils

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.KotlinFeature
import com.fasterxml.jackson.module.kotlin.KotlinModule
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.minturtle.careersupport.auth.utils.ApiTokenProvider
import org.minturtle.careersupport.common.exception.BadRequestException
import org.minturtle.careersupport.user.dto.UserInfoDto

class ApiTokenProviderTest {

    private val encryptUtils = EncryptUtils("X1XnwJ2Vrdw9wqdfX0rOdLfNJ8rwrvB9")

    private val objectMapper = ObjectMapper().apply {
        registerModule(
            KotlinModule.Builder()
                .configure(KotlinFeature.NullToEmptyCollection, true)
                .build()
        )
    }

    private val apiTokenProvider = ApiTokenProvider(encryptUtils, objectMapper)

    @Test
    fun `API 토큰을 생성할 수 있다`() {
        val testUserInfo = UserInfoDto("abc123", "nickname", "username")
        val token = apiTokenProvider.generate(testUserInfo)

        assertThat(token).isNotBlank()
        assertThat(token).startsWith("cs_")
        assertThat(token.length).isGreaterThan(3)
    }

    @Test
    fun `생성한 토큰을 해독해서 userId를 반환받을 수 있다`() {
        val testUserInfo = UserInfoDto("abc123", "nickname", "username")
        val token = apiTokenProvider.generate(testUserInfo)
        val userInfo = apiTokenProvider.decryptApiToken(token)

        assertThat(userInfo).isEqualTo(UserInfoDto("abc123", "nickname", "username"))
    }

    @ParameterizedTest
    @ValueSource(strings = ["", " ", "invalid", "cs_invalid"])
    fun `API 토큰이 잘못된 경우 BadRequestException을 throw한다`(invalidToken: String) {

        assertThatThrownBy {
            apiTokenProvider.decryptApiToken(invalidToken)
        }.isInstanceOf(BadRequestException::class.java)
    }
}